crates/apollo-compiler/src/validation/selection.rs

```diff
@@ -163,205 +275,274 @@ pub(crate) fn same_response_shape(
+    grouped_by_common_parents: OnceCell<Vec<Vec<FieldSelection<'doc>>>>,
+    same_response_shape_guard: OnceBool,
+    same_for_common_parents_guard: OnceBool,
+}
+
+impl<'doc> MergedFieldSet<'doc> {
+    fn new(selections: Vec<FieldSelection<'doc>>) -> Self {
+        Self {
+            selections,
+            grouped_by_output_names: Default::default(),
+            grouped_by_common_parents: Default::default(),
+            same_response_shape_guard: OnceBool::new(),
+            same_for_common_parents_guard: OnceBool::new(),
+        }
+    }
+
+    /// Given a set of fields, do all the fields that contribute to 1 output name have the same
+    /// shape?
+    ///
+    /// This prevents leaf output fields from having an inconsistent type.
+    fn same_response_shape_by_name(
+        &self,
+        validator: &mut FieldsInSetCanMerge<'_, 'doc>,
+        diagnostics: &mut DiagnosticList,
+    ) {
+        // No need to do this if this field set has been checked before.
+        if self.same_response_shape_guard.check() {
+            return;
+        }
+
+        for fields_for_name in self.group_by_output_name().values() {
+            let Some((field_a, rest)) = fields_for_name.split_first() else {
+                continue;
             };
-            let named_fragments = db.ast_named_fragments(file_id);
-            let mut merged_set = operation_fields(
-                &named_fragments,
-                &subfield_a_type.name,
-                &field_a.field.selection_set,
-            );
-            merged_set.extend(operation_fields(
-                &named_fragments,
-                &subfield_b_type.name,
-                &field_b.field.selection_set,
-            ));
-            let grouped_by_name = group_fields_by_name(merged_set);
-
-            for (_, fields_for_name) in grouped_by_name {
-                // 9. Given each pair of members subfieldA and subfieldB in fieldsForName:
-                let Some((subfield_a, rest)) = fields_for_name.split_first() else {
+            for (field_a, field_b) in std::iter::repeat(field_a).zip(rest.iter()) {
+                // Covers steps 3-5 of the spec algorithm.
+                if let Err(err) = same_output_type_shape(validator.schema, *field_a, *field_b) {
+                    diagnostics.push(field_b.field.location(), err);
                     continue;
-                };
-                for subfield_b in rest {
-                    // 9a. If SameResponseShape(subfieldA, subfieldB) is false, return false.
-                    same_response_shape(db, file_id, *subfield_a, *subfield_b)?;
                 }
             }
 
-            Ok(())
+            let mut nested_selection_sets = fields_for_name
+                .iter()
+                .map(|selection| &selection.field.selection_set)
+                .filter(|set| !set.selections.is_empty())
+                .peekable();
+            // TODO cache
```

Comment:
Is a follow-up planned or is this a "maybe some day" TODO? Or is the comment outdated and the relevant cache already there?

Reply:
Yeah this is outdated!